personhood/contract_popparty.go

```diff
@@ -0,0 +1,492 @@
+			if bytes.Compare(t.Tag, tag) == 0 {
+				return nil, nil, errors.New("this attendee already mined")
+			}
+		}
+		c.Miners = append(c.Miners, LRSTag{Tag: tag})
+
+		var coin byzcoin.Coin
+		var coinDarc darc.ID
+		coinAction := byzcoin.Update
+		coinIID := inst.Invoke.Args.Search("coinIID")
+		if coinIID == nil {
+			newDarcBuf := inst.Invoke.Args.Search("newDarc")
+			if newDarcBuf == nil {
+				return nil, nil, errors.New("need either coinIID or newDarc argument")
+			}
+			newDarc, err := darc.NewFromProtobuf(newDarcBuf)
+			if err != nil {
+				return nil, nil, errors.New("couldn't unmarshal darc: " + err.Error())
+			}
+			// Creating new darc for new user
+			log.Lvlf2("Creating new darc %x for user", newDarc.GetBaseID())
+			scs = append(scs, byzcoin.NewStateChange(byzcoin.Create,
+				byzcoin.NewInstanceID(newDarc.GetBaseID()), byzcoin.ContractDarcID,
+				newDarcBuf, darcID))
+			coinAction = byzcoin.Create
+			h := sha256.New()
+			h.Write([]byte("coin"))
+			h.Write(newDarc.GetBaseID())
+			coinIID = h.Sum(nil)
+			coinDarc = newDarc.GetBaseID()
+			log.Lvlf2("Creating new coin %x for user", coinIID)
+			coin.Name = byzcoin.NewInstanceID([]byte("SpawnerCoin"))
+		} else {
+			var cid string
+			var coinBuf []byte
+			coinBuf, _, cid, coinDarc, err = rst.GetValues(coinIID)
+			if cid != contracts.ContractCoinID {
+				return nil, nil, errors.New("coinIID is not a coin contract")
+			}
+			err = protobuf.Decode(coinBuf, &coin)
+			if err != nil {
+				return nil, nil, errors.New("couldn't unmarshal coin: " + err.Error())
+			}
+		}
+		err = coin.SafeAdd(c.MiningReward)
+		if err != nil {
+			return nil, nil, errors.New("couldn't add mining reward: " + err.Error())
+		}
+		coinBuf, err := protobuf.Encode(&coin)
+		if err != nil {
+			return nil, nil, errors.New("couldn't encode coin: " + err.Error())
+		}
+		scs = append(scs, byzcoin.NewStateChange(coinAction,
+			byzcoin.NewInstanceID(coinIID),
+			contracts.ContractCoinID, coinBuf, coinDarc))
+
+	default:
+		return nil, nil, errors.New("unknown command: " + inst.Invoke.Command)
+	}
+
+	// Storing new version of PopPartyStruct
+	ppiBuf, err := protobuf.Encode(&c.PopPartyStruct)
+	if err != nil {
+		return nil, nil, errors.New("couldn't marshal PopPartyStruct: " + err.Error())
+	}
+
+	// Update existing party structure
+	scs = append(scs, byzcoin.NewStateChange(byzcoin.Update, inst.InstanceID, ContractPopPartyID, ppiBuf, darcID))
+
+	return scs, coins, nil
+}
+
+// PopPartySpawn returns the instanceID of the newly created pop-party, or an error if it
+// wasn't successful.
+func PopPartySpawn(cl *byzcoin.Client, fs FinalStatement, dID darc.ID, reward uint64, signers ...darc.Signer) (popIID byzcoin.InstanceID, err error) {
+	var sigStrs []string
+	for _, sig := range signers {
+		sigStrs = append(sigStrs, sig.Identity().String())
+	}
+	signerCtrs, err := cl.GetSignerCounters(sigStrs...)
+	if err != nil {
+		return
+	}
+
+	fsBuf, err := protobuf.Encode(&fs)
```

Comment:
Pretty confused how it can be that this var of type FinalStatement is encoded then sent in via an arg called Description, the unmarshalled into c.Description. I can imagine how it might work, but there's either some code smell here, or maybe doc smell... or naming smell. (At least there's no caching to get wrong.)

Reply:
Thanks for catching this artifact - previously it was indeed a `FinalStatement` being passed in here. But when cleaning up it morphed into the `Description`.

Fun fact: it worked, because the `FinalStatement`'s first argument is a `Description`, so the `protobuf.Decode` produced a valid `Description` in the `Spawn`. Even though it looks really funny that it should've worked.